lib/gc/Dialect/OneDNNGraph/OneDNNGraphOps.cpp

```diff
@@ -181,22 +201,37 @@ LogicalResult onednn_graph::MatMulOp::inferReturnTypeComponents(
     MLIRContext *context, ::std::optional<Location> location,
     MatMulOp::Adaptor adaptor,
     SmallVectorImpl<ShapedTypeComponents> &inferredReturnShapes) {
-  // get batch dims from shape
-  auto extractBatch = [](const ShapeAdaptor &lhsShape,
-                         const ShapeAdaptor &rhsShape, int64_t range,
-                         int64_t diff, SmallVector<int64_t> &outDims) {
-    for (int64_t i = 0; i < range; i++) {
-      // TODO(longsheng): add OpTrait::util::getBroadcastedShape for batch
-      int64_t idx = i - diff;
-      if ((idx >= 0) && (lhsShape.getDimSize(i) != rhsShape.getDimSize(idx))) {
-        return failure();
-      }
-      outDims.push_back(lhsShape.getDimSize(i));
+  // get batch dims from 1 multi-batch mat shape
+  auto extractBatch = [](ShapedType shape, SmallVector<int64_t> &outDims) {
+    // assuming last 2 input dims are row and col
+    assert(shape.getRank() >= 2);
+    for (int64_t i = 0; i < shape.getRank() - 2; i++) {
+      outDims.push_back(shape.getDimSize(i));
     }
     return success();
   };
+  // get broadcasted batch dims from 2 multi-batch mat shape,
+  auto extractBroadcastBatch = [](ShapedType lhsType, ShapedType rhsType,
+                                  SmallVector<int64_t> &outDims) {
+    SmallVector<int64_t> lhsShape(lhsType.getShape());
+    SmallVector<int64_t> rhsShape(rhsType.getShape());
+    assert(lhsShape.size() >= 2 && rhsShape.size() >= 2);
+    // assuming last 2 input dims are row and col
```

Comment:
What guarantees it?

Reply:
In `else if (lRank > 1 && rRank > 1)`, it checks for both input rank >= 2, meaning 2 inputs are all matrix and may have batch dims.

Reply:
Transpose attr only `Controls whether to transpose the last two dimensions`, so batch dims always before last 2 dims according to the [onednn spec](https://spec.oneapi.io/onednn-graph/latest/ops/matrix/MatMul_1.html). If last two dimensions is somehow contain a batch dim, it is definitely wrong.